cpp/src/arrow/csv/writer.cc

```diff
@@ -318,9 +323,9 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
 
     std::vector<std::unique_ptr<ColumnPopulator>> populators(schema->num_fields());
     for (int col = 0; col < schema->num_fields(); col++) {
-      char end_char = col < schema->num_fields() - 1 ? ',' : '\n';
+      std::string end_chars = col < schema->num_fields() - 1 ? "," : options.eol;
```

Comment:
If you go the route of making the string a const reference in the populators you might need to move this `","` to a file-scoped static constant.